File: src/eko-core/common/xml.ts
// Use browser DOM APIs directly
import { Task, TaskNode, TaskTextNode } from '../types/core.types'
import type { PlannedTask, PlannedTaskTree } from '../types/core.types'

// All Workflow-related functions removed - Task-only architecture

export function parseTask(taskId: string, xml: string, done: boolean): Task | null {
  // Parse XML directly for single-agent task mode
  try {
    const parser = new DOMParser()
    const xmlToParse = xml.trim().startsWith('<root') ? xml : `<root>${xml}</root>`
    const doc = parser.parseFromString(xmlToParse, 'text/xml')
    const root = doc.documentElement

    if (root.tagName !== 'root') {
      return null
    }

    const nameElement = root.querySelector('name')
    const thoughtElement = root.querySelector('thought')
    const agentElement = root.querySelector('agent')

    const thought = thoughtElement?.textContent?.trim() || ''

    let description = ''
    let nodes: TaskNode[] = []
    let status: 'init' | 'running' | 'done' | 'error' = 'init'
    let agentXml = ''

    if (agentElement) {
      const taskElement = agentElement.querySelector('task')
      const taskText = taskElement?.textContent?.trim() || ''
      const name = nameElement?.textContent?.trim() || taskText || 'Untitled Task'
      description = taskText || name

      // Parse nodes from agent
      const nodeElements = agentElement.querySelectorAll('nodes > node, node')
      nodes = Array.from(nodeElements).map((nodeEl: Element): TaskNode => {
        const nodeText = nodeEl.textContent?.trim() || ''
        return {
          type: 'normal',
          text: nodeText,
        } as TaskTextNode
      })

      agentXml = agentElement.outerHTML
      status = done ? 'done' : 'init'
    } else {
      // Fallback: treat provided xml/name as description
      const nameText = nameElement?.textContent?.trim() || ''
      description = nameText || 'Untitled Task'
      agentXml = xml
    }

    // name fallback when agentElement missing
    const fallbackName = nameElement?.textContent?.trim() || description || 'Untitled Task'

    return {
      taskId,
      name: fallbackName,
      thought,
      description,
      nodes,
      status,
      xml: agentXml,
      modified: false,
      taskPrompt: undefined,
    }
  } catch (error) {
    console.error('Failed to parse task XML:', error)
    return null
  }
}

// Parse hierarchical task tree output (<root> with nested <subtasks><task>...)
export function parseTaskTree(xml: string): PlannedTaskTree | null {
  try {
    const parser = new DOMParser()
    const xmlToParse = xml.trim().startsWith('<root') ? xml : `<root>${xml}</root>`
    const doc = parser.parseFromString(xmlToParse, 'text/xml')
    const root = doc.documentElement
    if (root.tagName !== 'root') return null

    const parseNodes = (container: Element): { text: string }[] => {
      const nodesEl = container.querySelector(':scope > nodes')
      if (!nodesEl) return []
      const nodeEls = nodesEl.querySelectorAll(':scope > node')
      return Array.from(nodeEls)
        .map(el => ({ text: (el.textContent || '').trim() }))
        .filter(n => n.text)
    }

    const parsePlannedTask = (taskEl: Element): PlannedTask => {
      const name = taskEl.querySelector(':scope > name')?.textContent?.trim()
      const desc = taskEl.querySelector(':scope > task')?.textContent?.trim()
      const nodes = parseNodes(taskEl)
      const subtasks: PlannedTask[] = []
      const subsEl = taskEl.querySelector(':scope > subtasks')
      if (subsEl) {
        const taskChildren = subsEl.querySelectorAll(':scope > task')
        taskChildren.forEach(child => {
          subtasks.push(parsePlannedTask(child))
        })
      }
      return {
        name,
        description: desc,
        nodes,
        subtasks: subtasks.length > 0 ? subtasks : undefined,
      }
    }

    // root level
    const name = root.querySelector(':scope > name')?.textContent?.trim()
    const thought = root.querySelector(':scope > thought')?.textContent?.trim()
    const desc = root.querySelector(':scope > task')?.textContent?.trim()
    const nodes = parseNodes(root)
    const subtasks: PlannedTask[] = []
    const subsEl = root.querySelector(':scope > subtasks')
    if (subsEl) {
      subsEl.querySelectorAll(':scope > task').forEach(t => subtasks.push(parsePlannedTask(t)))
    }

    const tree: PlannedTaskTree = {
      name,
      thought,
      description: desc,
      nodes,
      subtasks: subtasks.length > 0 ? subtasks : undefined,
    }
    return tree
  } catch (e) {
    console.error('Failed to parse task tree XML:', e)
    return null
  }
}

// Build minimal <agent> XML string from planned task data
export function buildAgentXmlFromPlanned(planned: PlannedTask): string {
  try {
    const doc = new DOMParser().parseFromString('<agent></agent>', 'text/xml')
    const agent = doc.documentElement
    const t = doc.createElement('task')
    t.textContent = (planned.description || planned.name || '').trim()
    agent.appendChild(t)
    if (planned.nodes && planned.nodes.length > 0) {
      const nodesEl = doc.createElement('nodes')
      for (const n of planned.nodes) {
        const nodeEl = doc.createElement('node')
        nodeEl.textContent = (n.text || '').trim()
        nodesEl.appendChild(nodeEl)
      }
      agent.appendChild(nodesEl)
    }
    return agent.outerHTML || '<agent />'
  } catch {
    return '<agent />'
  }
}

export function resetTaskXml(task: Task) {
  task.modified = false
}

export function buildAgentRootXml(
  agentXml: string,
  taskPrompt: string,
  nodeCallback?: (nodeId: string, node: Element) => void
): string {
  try {
    const parser = new DOMParser()
    const doc = parser.parseFromString(`<root>${agentXml}</root>`, 'text/xml')
    const root = doc.documentElement

    // Insert user's instruction to ensure LLM gets the actual task context
    if (taskPrompt && taskPrompt.trim().length > 0) {
      const instructionEl = doc.createElement('instruction')
      // Use textContent to avoid breaking XML with special chars
      instructionEl.textContent = taskPrompt
      // Prepend instruction before nodes for visibility
      if (root.firstChild) {
        root.insertBefore(instructionEl, root.firstChild)
      } else {
        root.appendChild(instructionEl)
      }
    }

    // Apply node callback if provided
    if (nodeCallback) {
      const nodes = root.querySelectorAll('node')
      for (let i = 0; i < nodes.length; i++) {
        const node = nodes[i] as Element
        nodeCallback(`node-${i}`, node)
      }
    }

    return root.outerHTML || root.toString()
  } catch (error) {
    console.error('Failed to build agent root XML:', error)
    return agentXml
  }
}

export function getOuterXML(node: Element): string {
  return node.outerHTML || node.toString()
}

export function extractAgentXmlNode(agentXml: string, nodeSelector: string): Element | null {
  try {
    const parser = new DOMParser()
    const doc = parser.parseFromString(`<root>${agentXml}</root>`, 'text/xml')
    const node = doc.querySelector(nodeSelector)
    return node as Element | null
  } catch (error) {
    console.error('Failed to extract agent XML node:', error)
    return null
  }
}

File: src-tauri/src/agent/common/xml.rs
use std::io::Cursor;

use crate::agent::error::{AgentError, AgentResult};
use crate::agent::types::{PlannedTask, PlannedTaskNode, TaskDetail, TaskDetailStatus, TaskNode};
use xmltree::{Element, XMLNode};

fn ensure_root(xml: &str) -> String {
    let trimmed = xml.trim();
    if trimmed.starts_with("<root") {
        trimmed.to_string()
    } else {
        format!("<root>{}</root>", trimmed)
    }
}

pub fn parse_task_detail(task_id: &str, xml: &str, done: bool) -> AgentResult<TaskDetail> {
    let wrapped = ensure_root(xml);
    let root = Element::parse(Cursor::new(wrapped.as_bytes()))?;

    if root.name != "root" {
        return Err(AgentError::XmlParse(
            "Task XML must have <root> element".to_string(),
        ));
    }

    let name_text = root
        .get_child("name")
        .and_then(|el| el.get_text())
        .map(|v| v.to_string())
        .unwrap_or_default();
    let thought_text = root
        .get_child("thought")
        .and_then(|el| el.get_text())
        .map(|v| v.to_string())
        .unwrap_or_default();

    let (description, nodes, agent_xml) = if let Some(agent_el) = root.get_child("agent") {
        let description = agent_el
            .get_child("task")
            .and_then(|el| el.get_text())
            .map(|v| v.to_string())
            .unwrap_or_else(|| name_text.clone());

        let nodes = collect_task_nodes(agent_el.get_child("nodes"));

        let mut buf = Vec::new();
        agent_el.write(Cursor::new(&mut buf))?;
        let agent_xml = String::from_utf8(buf).unwrap_or_default();
        (description, nodes, agent_xml)
    } else {
        (name_text.clone(), Vec::new(), xml.to_string())
    };

    Ok(TaskDetail {
        task_id: task_id.to_string(),
        name: if name_text.is_empty() {
            description.clone()
        } else {
            name_text
        },
        thought: thought_text,
        description,
        nodes,
        status: if done {
            TaskDetailStatus::Done
        } else {
            TaskDetailStatus::Init
        },
        xml: agent_xml,
        modified: false,
        task_prompt: None,
    })
}

fn collect_task_nodes(nodes_el: Option<&Element>) -> Vec<TaskNode> {
    let mut nodes = Vec::new();
    if let Some(el) = nodes_el {
        for child in &el.children {
            if let XMLNode::Element(node_el) = child {
                if node_el.name == "node" {
                    let text = node_el
                        .get_text()
                        .map(|v| v.to_string())
                        .unwrap_or_default();
                    nodes.push(TaskNode { text });
                }
            }
        }
    }
    nodes
}

pub fn parse_task_tree(xml: &str) -> AgentResult<PlannedTask> {
    let wrapped = ensure_root(xml);
    let root = Element::parse(Cursor::new(wrapped.as_bytes()))?;
    if root.name != "root" {
        return Err(AgentError::XmlParse(
            "Tree XML must have <root> element".to_string(),
        ));
    }

    Ok(parse_planned_task(&root))
}

fn parse_planned_task(element: &Element) -> PlannedTask {
    let mut planned = PlannedTask::default();
    planned.name = element
        .get_child("name")
        .and_then(|el| el.get_text())
        .map(|v| v.to_string());
    planned.thought = element
        .get_child("thought")
        .and_then(|el| el.get_text())
        .map(|v| v.to_string());
    planned.description = element
        .get_child("task")
        .and_then(|el| el.get_text())
        .map(|v| v.to_string());

    planned.nodes = element.get_child("nodes").map(|nodes_el| {
        nodes_el
            .children
            .iter()
            .filter_map(|child| match child {
                XMLNode::Element(el) if el.name == "node" => {
                    el.get_text().map(|v| PlannedTaskNode {
                        text: v.to_string(),
                    })
                }
                _ => None,
            })
            .collect()
    });

    if let Some(subtasks_el) = element.get_child("subtasks") {
        let subtasks = subtasks_el
            .children
            .iter()
            .filter_map(|child| match child {
                XMLNode::Element(el) if el.name == "task" => Some(parse_planned_task(el)),
                _ => None,
            })
            .collect::<Vec<_>>();
        if !subtasks.is_empty() {
            planned.subtasks = Some(subtasks);
        }
    }

    planned
}

pub fn build_agent_xml_from_planned(planned: &PlannedTask) -> AgentResult<String> {
    let mut agent_el = Element::new("agent");

    if let Some(description) = planned.description.as_ref().or(planned.name.as_ref()) {
        let mut task_el = Element::new("task");
        task_el
            .children
            .push(XMLNode::Text(description.to_string()));
        agent_el.children.push(XMLNode::Element(task_el));
    }

    if let Some(nodes) = &planned.nodes {
        if !nodes.is_empty() {
            let mut nodes_el = Element::new("nodes");
            for node in nodes {
                let mut node_el = Element::new("node");
                node_el.children.push(XMLNode::Text(node.text.to_string()));
                nodes_el.children.push(XMLNode::Element(node_el));
            }
            agent_el.children.push(XMLNode::Element(nodes_el));
        }
    }

    let mut buf = Vec::new();
    agent_el.write(Cursor::new(&mut buf))?;
    String::from_utf8(buf).map_err(AgentError::Utf8)
}
